### PR TITLE
🎓 Topic control logic --- DIG file

### PR DESCRIPTION
**Pull request by jameshughes89**
### What

Create the instruction register and control logic 
A hex file for the look up table's control logic patterns. 
A script to make the control logic patterns. 

image

### How

Instruction register is a register that can input from the data bus, but output only the lease significant 4 bits. The most significant 4 bits serve as the operator/instruction and the least significant are the operand, where applicable. 

The most significant 4 bits refers to one of 16 possible instructions. Then there is a microcode 2 bit counter that steps through the 4 possible microcodes for each of the 16 instructions. Originally I had up to 4 microcodes, but I simplified things. 

I may also need to invert the clock into the microcode counter such that it steps _between_ clock cycles. This is unclear and I will figure that out once I hook it up to the rest of the system. It may be fine since the simulator is idealized. In fact, Idid notice some funny race conditions when it was inverted, so who knows how it will play out. 

### Testing

I split the tests into 2 sets. One checks the instruction register, the other checks the control logic. This (a) keeps things separated, and (b) keeps the number of columns in the test cases lower and easier to mentally parse. 

image
image

NOTE: I checked literally all 64 microcodes in the control logic tests. 

I am thinking that I will modify the instruction set as the course progresses, meaning the control logic may need to change quite a bit. However, if this happens, I think this schematic will remain unchanged, but a NEW schematic will be made with any necessary changes and updated tests.

### Additional Notes

Maybe I should split this into 2 designs? One for the register, then one for the control logic lookup? It's probably fine as is at this stage of the semester.

### CODE REVIEW

File: site/topics/control-logic/create_control_logic_patterns_for_look_up_table.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  - PCI (Program Counter In)
+  - PCO (Program Counter Out)
+  - ARI (A Register In)
+  - ARO (A Register Out)
+  - BRI (B Register In)
+  - BRO (B Register Out)
+  - ALU (ALU Out)
+  - SUB (Subtraction)
+  - ORI (Output Register In)
+  - SGN (Signed Integer)
```
Comment (jameshughes89) [inline site/topics/control-logic/create_control_logic_patterns_for_look_up_table.py]: im thinking the order of these will be reversed to help with the layout of the system, but I will hold off on that until I get things together to see if it is necessary

**Comment by jameshughes89** [comment]
Made draft as I will be changing the order of the control logic a little to fit nicer into the whole system.

**Comment by jameshughes89** [comment]
Here is how the layout is when adding the control logic to the system. 

image

Order ends up moving around a lot. See example below. 

image

**Comment by jameshughes89** [comment]
Halt would need to be added
image

**Comment by jameshughes89** [comment]
Final idea
image

**Comment by jameshughes89** [comment]
Ready for review. 

1. Reordered the control logic lines to fit into the whole system a little better
2. Update the script for making the hex file the look up table uses
3. Although not necessary, I reordered the tests to align with the new ordering of the control lines

**Review by twentylemon — APPROVED** [review]
he's yoloing

**Comment by jameshughes89** [comment]
> he's yoloing

I feel (hope?) this is the most complex part of the course. Fortunately it will be explained in pieces within the topics, but the module itself is a big boy.

**Pull request merged**